### PR TITLE
Benj9250/mesh cooking

### PR DESCRIPTION
**Pull request by bneukom**
Added mesh cooking. To properly support mesh cooking we also added better support for instance caching beyond just one initial shape. This also should help performance as we don't have to encode each instanced geometry assed for each individual initial shape.

### CODE REVIEW

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioEditorModule.cpp [hunk 1]
```diff
@@ -15,28 +15,40 @@
...
 namespace
 {
 
 bool IsViableVitruvioActor(AActor* Actor)
 {
+
+	if (Cast<AVitruvioActor>(Actor))
+	{
+		return false;
+	}
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioEditorModule.cpp:45]: The naming of this function is a bit confusing. Should either be commented or renamed.

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+	// Regenerate the selected Actors to make sure we have a model to cook.
+	for (AActor* Actor : Actors)
+	{
+		UVitruvioComponent* VitruvioComponent = Actor->FindComponentByClass<UVitruvioComponent>();
+		if (!VitruvioComponent)
+		{
+			continue;
+		}
+
+		VitruvioComponent->Generate();
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:330]: I assume there is a reason why everything needs to be regenerated. Is it not possible to find out if the model has already been generated with the current configuration?

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+			FStaticMeshCache MeshCache;
+
+			for (AActor* Actor : Actors)
+			{
+				CookTask.EnterProgressFrame(1);
+
+				AActor* OldAttachParent = Actor->GetAttachParentActor();
+
+				// Spawn new Actor with persisted geometry
+				AActor* CookedActor = Actor->GetWorld()->SpawnActor<AActor>(Actor->GetActorLocation(), Actor->GetActorRotation());
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:258]: Is it possible to name the actor the same as the old one (or similar)?

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+							InstancedStaticMeshComponent->AddInstance(Transform);
+						}
+					}
+				}
+
+				// Destroy the old procedural Vitruvio Actor
+				Actor->Destroy();
+			}
+
+			IsCooking = false;
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:315]: Probably doesn't matter, but I would move this line below `ModelsGeneratedHandle.Reset();`

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+	OriginalMip.BulkData.Unlock();
+
+	NewTexture->PostEditChange();
+	TexturePackage->MarkPackageDirty();
+	FAssetRegistryModule::AssetCreated(NewTexture);
+
+	TextureCache.Add(Original, NewTexture);
+
+	return NewTexture;
+}
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:106]: Would it be possible to replace this code with a generic DuplicateAsset call (e.g. SequenceRecorderUtils.h::DuplicateAsset)?

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+	PersistedMesh->PostEditChange();
+	PersistedMesh->MarkPackageDirty();
+
+	// Notify asset registry of new asset
+	FAssetRegistryModule::AssetCreated(PersistedMesh);
+
+	MeshCache.Add(Mesh, PersistedMesh);
+	return PersistedMesh;
+}
+} // namespace
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:217]: Same here but the textures/materials would need to be replaced after the duplicate call.

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+	const FString MaterialName = Material->GetName();
+	const FString MaterialPackagePath = FPaths::Combine(Path, TEXT("Materials"), MaterialName);
+	UPackage* MaterialPackage = CreatePackage(*MaterialPackagePath);
+
+	UMaterialInstanceConstantFactoryNew* MaterialFactory = NewObject<UMaterialInstanceConstantFactoryNew>();
+	MaterialFactory->InitialParent = Material->Parent;
+
+	UMaterialInstanceConstant* NewMaterial = Cast<UMaterialInstanceConstant>(MaterialFactory->FactoryCreateNew(
+		UMaterialInstanceConstant::StaticClass(), MaterialPackage, *MaterialName, RF_Public | RF_Standalone, nullptr, GWarn));
+	FAssetRegistryModule::AssetCreated(NewMaterial);
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:125]: Is there a reason why this call is placed here and not at the end of the function? (for the texture/mesh its after `PostEditChange/MarkPackageDirty`

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioEditorModule.cpp [hunk 8]
```diff
@@ -152,6 +172,15 @@ TSharedRef<FExtender> ExtendLevelViewportContextMenuForVitruvioComponents(const
 					FSlateIcon(), AddVitruvioComponentAction);
 			}
 
+			if (HasAnyVitruvioActor(SelectedActors))
+			{
+				const FUIAction CookVitruvioActorsAction(FExecuteAction::CreateStatic(CookVitruvioActors, SelectedActors));
+
+				MenuBuilder.AddMenuEntry(FText::FromString("Cook Vitruvio Actor"),
```
Comment (bstegerESRI) [inline VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioEditorModule.cpp:179]: We should maybe choose a name which is a bit easier to understand (e.g. "Convert to static mesh)

**Review by bstegerESRI — APPROVED** [review]
Looks good & works! Only found some minor stuff which could be changed.

File: VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioMesh.cpp [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+		return MaterialCache[MaterialAttributes];
+	}
+	else
+	{
+		const FName UniqueMaterialName(Name);
+		UMaterialInstanceDynamic* Material = Vitruvio::GameThread_CreateMaterialInstance(Outer, UniqueMaterialName, OpaqueParent, MaskedParent,
+																						 TranslucentParent, MaterialAttributes, TextureCache);
+		MaterialCache.Add(MaterialAttributes, Material);
+		return Material;
+	}
```
[the remark this answers is not in the document]
Comment (bneukom) [reply VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioMesh.cpp:22]: Should only be run on the game thread. But good point, I'll add a check(IsInGameThread()) for that.

File: VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/MeshCache.cpp [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	return Mesh;
+}
+
+void FMeshCache::Invalidate()
+{
+	FScopeLock Lock(&MeshCacheCriticalSection);
+	for (auto& Entry : Cache)
+	{
+		Entry.Value->Invalidate();
+	}
```
[the remark this answers is not in the document]
Comment (bneukom) [reply VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/MeshCache.cpp:30]: Makes sense!

File: VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Public/VitruvioModule.h [hunk 11]
```diff
@@ -230,6 +230,7 @@ class VitruvioModule final : public IModuleInterface, public FGCObject
 	mutable TMap<TLazyObjectPtr<URulePackage>, FGraphEventRef> ResolveMapEventGraphRefCache;
 
 	mutable FCriticalSection LoadResolveMapLock;
+	mutable FCriticalSection MeshCacheCriticalSection;
```
[the remark this answers is not in the document]
Comment (bneukom) [reply VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Public/VitruvioModule.h:233]: True, deleted :)

File: VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioMesh.cpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+#include "MaterialConversion.h"
+#include "StaticMeshAttributes.h"
+
+UMaterialInstanceDynamic* CacheMaterial(UMaterial* OpaqueParent, UMaterial* MaskedParent, UMaterial* TranslucentParent,
+										TMap<FString, Vitruvio::FTextureData>& TextureCache,
+										TMap<Vitruvio::FMaterialAttributeContainer, UMaterialInstanceDynamic*>& MaterialCache,
+										const Vitruvio::FMaterialAttributeContainer& MaterialAttributes, const FName& Name, UObject* Outer)
+{
+	if (MaterialCache.Contains(MaterialAttributes))
```
Comment (mistafunk) [inline VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioMesh.cpp:11]: AMaybe you want to work with an iterator to avoid duplicate lookup?

Comment (bneukom) [reply VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/VitruvioMesh.cpp:11]: Good point @mistafunk! Will do.

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp (hunk 2 above)
Comment (bneukom) [reply VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:341]: Yes we could theoretically save that state somehwere. But probably more something for a future ticket.

File: VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp (hunk 3 above)
Comment (bneukom) [reply VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp:269]: Makes sense, will do.

**Pull request merged**